sql: table DROP in chunks

vivekmenezes opened the pull request:
A table DROP in chunks so that the kv layer doesn't
need to manage large table deletions.

Added a test for interleaved table deletions.

table TRUNCATE is still not chunked.
#7499, #2003

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8885)

<!-- Reviewable:end -->

vivekmenezes committed:
sql: table DROP in chunks

A table DROP in chunks so that the kv layer doesn't
need to manage large table deletions.

Added a test for interleaved table deletions.

table TRUNCATE is still not chunked.

tschottdorf commented:
> The use of a single transaction
will not cause conflicts during the table DROP case, because once a table
is marked for deletion it cannot be read/written to. Conflicts will be
observed during the table TRUNCATE case.

That's not quite true, the use of a single transaction will still cause problems if the transaction fails at any point. The "next" transaction trying to do the same drop will have to start from the beginning and run into all of the intents one by one. To fix the underlying problem, use multiple transactions and commit each chunk individually. That way, the most you can run into is one chunk of intents.

vivekmenezes commented:
True. I was trying to use the same code for truncation and deletion.  I think we fundamentally need to change TRUNCATE (assign a new descriptor ID for the table) before we can move forward with this change.  Hold off on reviewing this change.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-:-KQLvf8CKMx7NUVB0mJU:b-z4ey6w)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
I think it's a good idea to implement truncate by changing the table id and deleting the data in the background. I have a small concern that we only get 59 = (encoding.intSmall - keys.MaxReservedDescID) table ids before they take 2 bytes to represent. It will possibly be worth it later to reuse ids of tables we've truncated once it's safe to. Can you file an issue?

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-:-KQLzEvBZpF8hdJpW0K7:b3g45m9)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
This PR is now only limited to chunking table DROP. I've left table TRUNCATE as is and will follow up with a fix for that later. It is ready for review. Thanks

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-:-KQMc7jvyM71CjQeMtdW:bsp54mb)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
@paperstreet there is already an existing issue for optimizing TRUNCATE #2003

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-:-KQMjOZmJCVlKFMs8v85:b-8ekgx2)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
@paperstreet any thoughts on this change? Thanks

paperstreet commented:
Mostly looks okay, but these tests have gotten out of hand and I'm having trouble spotting the differences between them. I think they need to be DRY'd before this PR lands and it gets any worse.

FWIW, that issue doesn't mention anything about table id encoding length and reusing them

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[sql/drop.go, line 605 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQSJrNrHAeWsZXxmA-6:-KQSJrNrHAeWsZXxmA-7:b-19vh1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2d042b2964340d1e8f92dd38058a901b0d77b7/sql/drop.go#L605)):*
> ```Go
> // therefore the entire table can be deleted with no concern for conflicts (we
> // can even eliminate the need to use a transaction within
> // truncateTableInChunks at a later stage if it proves inefficient).
> ```

mentioning `truncateTableInChunks` is just asking for rot. I think "within the chunks" or similar is just as informational

---

*[sql/drop.go, line 607 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQSIAephcvjAY1AibiE:-KQSIAephcvjAY1AibiF:b-1qz0xb) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2d042b2964340d1e8f92dd38058a901b0d77b7/sql/drop.go#L607)):*
> ```Go
> // truncateTableInChunks at a later stage if it proves inefficient).
> func truncateAndDropTable(tableDesc *sqlbase.TableDescriptor, db *client.DB) error {
> 	// Truncate table data in chunks.
> ```

Unnecessary

---

*[sql/drop.go, line 612 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQSIClvI4i9wRMh-xZB:-KQSIClvI4i9wRMh-xZC:bx8n95t) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2d042b2964340d1e8f92dd38058a901b0d77b7/sql/drop.go#L612)):*
> ```Go
> 	}
> 
> 	// Delete the table meta data.
> ```

Also probably unnecessary.

---

*[sql/drop_test.go, line 344 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQSITyyOZ6ktdvyWT__:-KQSITyyOZ6ktdvyWT_a:b-1qz0xb) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2d042b2964340d1e8f92dd38058a901b0d77b7/sql/drop_test.go#L344)):*
> ```Go
> 	}
> 
> 	// Bulk insert.
> ```

Unnecessary

---

*[sql/drop_test.go, line 405 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQSEywxX6AzOz4qy_hi:-KQSEywxX6AzOz4qy_hj:b1dhgtx) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2d042b2964340d1e8f92dd38058a901b0d77b7/sql/drop_test.go#L405)):*
> ```Go
> // TestDropTableInterleaved tests dropping a table that is interleaved within
> // another table.
> func TestDropTableInterleaved(t *testing.T) {
> ```

these tests badly need to be DRY'd

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-:-KQSEA0QP5eAzG4ZMB7T:bsc3ku1)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Okay ready for another review. Thanks!

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[sql/drop.go, line 605 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQSJrNrHAeWsZXxmA-6:-KQVmIcFpcQceXpO-lzS:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2d042b2964340d1e8f92dd38058a901b0d77b7/sql/drop.go#L605)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> mentioning `truncateTableInChunks` is just asking for rot. I think "within the chunks" or similar is just as informational
</details>
done

---

*[sql/drop.go, line 607 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQSIAephcvjAY1AibiE:-KQVnLiGKjxwSfzqi2R3:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2d042b2964340d1e8f92dd38058a901b0d77b7/sql/drop.go#L607)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Unnecessary
</details>
done

---

*[sql/drop.go, line 612 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQSIClvI4i9wRMh-xZB:-KQVnVq1-6GHAILxsusZ:b-o69pmv) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2d042b2964340d1e8f92dd38058a901b0d77b7/sql/drop.go#L612)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Also probably unnecessary.
</details>
changed to

// Finished deleting all the table data, now delete the table meta data.

---

*[sql/drop_test.go, line 344 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQSITyyOZ6ktdvyWT__:-KQWBt3PBVmkMVtJiEZx:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2d042b2964340d1e8f92dd38058a901b0d77b7/sql/drop_test.go#L344)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Unnecessary
</details>
removed

---

*[sql/drop_test.go, line 405 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQSEywxX6AzOz4qy_hi:-KQWBqvq5Y7zWRuOJQBF:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd2d042b2964340d1e8f92dd38058a901b0d77b7/sql/drop_test.go#L405)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> these tests badly need to be DRY'd
</details>
fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-:-KQWBuGNDugK0kWWG06o:b-5xvcp6)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[sql/drop.go, line 606 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQWGnyc3AmCu8y4f-Kg:-KQWGnyc3AmCu8y4f-Kh:b-o5sbi) ([raw file](https://github.com/cockroachdb/cockroach/blob/88f95a439281f7fe8fbd5f279425d9d8d7591aee/sql/drop.go#L606)):*
> ```Go
> // can even eliminate the need to use a transaction for each chunk at a later
> // stage if it proves inefficient).
> func truncateAndDropTable(tableDesc *sqlbase.TableDescriptor, db *client.DB) error {
> ```

as long as you're here, mind making this function take a `Context` and change the `Context.TODO()` below?

---

*[sql/tablewriter.go, line 587 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQWI2RVpsFlTC8JUWGc:-KQWI2RWL_I6gDlk4nqk:b5yva7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/88f95a439281f7fe8fbd5f279425d9d8d7591aee/sql/tablewriter.go#L587)):*
> ```Go
> // resume is the resume-span which should be used for the table deletion when
> // the table deletion is chunked. The first call to this method should use a
> // zero resume-span. After a chunk is deleted a new resume-span is returned.
> ```

how come it's a "resume-span" and not a "resume-key" ?

---

*[sql/tablewriter.go, line 589 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQWJz4zZ6RLQvAWEcLf:-KQWJz4zZ6RLQvAWEcLg:b-b05z6f) ([raw file](https://github.com/cockroachdb/cockroach/blob/88f95a439281f7fe8fbd5f279425d9d8d7591aee/sql/tablewriter.go#L589)):*
> ```Go
> // zero resume-span. After a chunk is deleted a new resume-span is returned.
> //
> // limit is a limit on either the number of keys or table-rows (for
> ```

This comment confused me before I found out what's going on. At least I think we should expand the comment and explain the difference, and how a scan is not needed if the table is not interleaved (either as the parent or the child). 
But better yet, I think the difficulty in defining what it means is a sign that maybe the higher-layer should not be supplying it. Why not let this function do the chunking?
And same for index deletion functions. Unless we had this conversation then too and I forgot.

---

*[sql/tablewriter.go, line 619 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQWMm_sqwiPJso61niu:-KQWMm_sqwiPJso61niv:b-4hbagk) ([raw file](https://github.com/cockroachdb/cockroach/blob/88f95a439281f7fe8fbd5f279425d9d8d7591aee/sql/tablewriter.go#L619)):*
> ```Go
> 		log.Infof(ctx, "DelRange %s - %s", resume.Key, resume.EndKey)
> 	}
> 	td.b.DelRange(resume.Key, resume.EndKey, false)
> ```

please comment with `/* returnKeys */`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-:-KQWNEZ1NxIwwzh_18UH:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 4 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[sql/drop.go, line 606 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQWGnyc3AmCu8y4f-Kg:-KQX7YNojor3YV3uJG_O:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/88f95a439281f7fe8fbd5f279425d9d8d7591aee/sql/drop.go#L606)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> as long as you're here, mind making this function take a `Context` and change the `Context.TODO()` below?
</details>
done

---

*[sql/tablewriter.go, line 587 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQWI2RVpsFlTC8JUWGc:-KQX7c0AGa94AEZHGcxI:b-151nf9) ([raw file](https://github.com/cockroachdb/cockroach/blob/88f95a439281f7fe8fbd5f279425d9d8d7591aee/sql/tablewriter.go#L587)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> how come it's a "resume-span" and not a "resume-key" ?
</details>
resume-span is what is exported from the kv layer. The sql layer should just use that and not care what it contains.

---

*[sql/tablewriter.go, line 589 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQWJz4zZ6RLQvAWEcLf:-KQX7xW2Yqb0FQDBcJX_:b-zdr2lf) ([raw file](https://github.com/cockroachdb/cockroach/blob/88f95a439281f7fe8fbd5f279425d9d8d7591aee/sql/tablewriter.go#L589)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> This comment confused me before I found out what's going on. At least I think we should expand the comment and explain the difference, and how a scan is not needed if the table is not interleaved (either as the parent or the child). 
> But better yet, I think the difficulty in defining what it means is a sign that maybe the higher-layer should not be supplying it. Why not let this function do the chunking?
> And same for index deletion functions. Unless we had this conversation then too and I forgot.
</details>
For the index deletion case each index entry is just one key, so the scan and non scan approach are consistent in the parameter referring to keys.

In this case, you either are deleting keys, or deleting sql rows. I think I'm going to leave it as is for now because I prefer passing the chunking parameter in from above. If in the future we need two separate parameters I'll split this up into two functions and pass in different parameters from above, or take the parameter look up the number of columns/indexes and better approximate the number of keys per row for the table and use that.

---

*[sql/tablewriter.go, line 619 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885#-KQWMm_sqwiPJso61niu:-KQX9AbL9tc8rn5lkjV8:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/88f95a439281f7fe8fbd5f279425d9d8d7591aee/sql/tablewriter.go#L619)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> please comment with `/* returnKeys */`
</details>
done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8885)*
<!-- Sent from Reviewable.io -->
